sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaIO.java

```diff
@@ -335,22 +346,38 @@
      * of how the partitions are distributed among the splits.
      */
     public Read<K, V> withTopic(String topic) {
-      return withTopics(ImmutableList.of(topic));
+      return withTopic(ValueProvider.StaticValueProvider.of(topic));
     }
 
     /**
-     * Sets a list of topics to read from. All the partitions from each
-     * of the topics are read.
+     * Like above but with a {@link ValueProvider ValueProvider&lt;String&gt;}.
+     */
+    public Read<K, V> withTopic(ValueProvider<String> topic) {
+      return withTopics(topic);
+    }
+
+    /**
+     * Sets a list of topics to read from with a {@link List List&lt;String&gt;}.
+     * All the partitions from each of the topics are read.
      *
      * <p>See {@link KafkaUnboundedSource#split(int, PipelineOptions)} for description
      * of how the partitions are distributed among the splits.
      */
     public Read<K, V> withTopics(List<String> topics) {
-      checkState(
-          getTopicPartitions().isEmpty(), "Only topics or topicPartitions can be set, not both");
-      return toBuilder().setTopics(ImmutableList.copyOf(topics)).build();
+      return withTopics(ValueProvider.StaticValueProvider.of(Joiner.on(',').join(topics)));
     }
 
+    /**
+     * Like above but with a {@link ValueProvider ValueProvider&lt;String&gt;}.
+     * The format is comma separated String of topics
+     */
+    public Read<K, V> withTopics(ValueProvider<String> topics) {
+      checkState(
+              getTopicPartitions().get().isEmpty(),
```

Comment:
Calling `.get()` here means that this can not be used with a runtime value provider at all which defeats the purpose of this change. `.get()` can only be called inside code that runs while the pipeline is running, but not at construction time - i.e. only inside code of `DoFn`s, effectively.

Reply:
I have changed the getTopicPartitions().get().isEmpty() to getTopicPartitions() == null